keras_nlp/models/llama/llama_attention.py

```diff
@@ -0,0 +1,197 @@
+
+        self.num_key_value_groups = num_query_heads // num_key_value_heads
+        self.rotary_max_wavelength = rotary_max_wavelength
+
+        self._kernel_initializer = keras.initializers.get(kernel_initializer)
+        self._bias_initializer = keras.initializers.get(bias_initializer)
+        self.max_sequence_length = max_sequence_length
+
+        self.rope_scaling_type = rope_scaling_type
+        self.rope_scaling_factor = rope_scaling_factor
+
+        if rope_scaling_type is None:
+            self.rotary_embedding_layer = RotaryEmbedding(
+                max_wavelength=rotary_max_wavelength
+            )
+        else:
+            if rope_scaling_type == "linear":
+                self.rotary_embedding_layer = RotaryEmbedding(
+                    max_wavelength=rotary_max_wavelength,
+                    scaling_factor=rope_scaling_factor,
+                )
+            else:
+                raise ValueError(
+                    f"Unknown RoPE scaling type {rope_scaling_type}"
+                )
+
+    def build(self, inputs_shape):
+        self.hidden_dim = inputs_shape[-1]
+        self.attn_head_size = self.hidden_dim // self.num_query_heads
+
+        self._query_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_query_heads, self.attn_head_size),
+            **self._get_common_kwargs_for_sublayer(use_bias=False),
+            name="query",
+        )
+        self._query_dense.build(inputs_shape)
+
+        self._key_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_key_value_heads, self.attn_head_size),
+            **self._get_common_kwargs_for_sublayer(use_bias=False),
+            name="key",
+        )
+        self._key_dense.build(inputs_shape)
+
+        self._value_dense = keras.layers.EinsumDense(
+            equation="abc,cde->abde",
+            output_shape=(None, self.num_key_value_heads, self.attn_head_size),
+            **self._get_common_kwargs_for_sublayer(use_bias=False),
+            name="value",
+        )
+        self._value_dense.build(inputs_shape)
+
+        self._softmax = keras.layers.Softmax(axis=-1, name="attention_softmax")
+
+        self._output_dense = keras.layers.EinsumDense(
+            equation="abc,cd->abd",
+            output_shape=(None, self.hidden_dim),
+            **self._get_common_kwargs_for_sublayer(use_bias=False),
+            name="attention_output",
+        )
+        self._output_dense.build(inputs_shape)
+
+        self.built = True
+
+    def call(
+        self, hidden_states, attention_mask=None, return_attention_scores=False
+    ):
+        query = self._query_dense(hidden_states)
+        key = self._key_dense(hidden_states)
+        value = self._value_dense(hidden_states)
+
+        query = self.rotary_embedding_layer(query)
+        key = self.rotary_embedding_layer(key)
```

Comment:
The PyTorch implementation uses `torch.view_as_complex`/`torch.view_as_real` to convert to/from complex numbers while we just `ops.split`/`ops.concatenate` the input in the rotary embeddings layer. To make both the operations numerically equivalent, I think we need to transform the input before feeding them to the rotary embedding layer and then reverse transform the output. Something like:

```python
        query = ops.concatenate([query[... ::2], query[..., 1::2]], axis=-1)
        query = self.rotary_embedding_layer(query)
        query = ops.reshape(ops.stack(ops.split(query, 2, axis=-1), axis=-1), query.shape)
        key = self.rotary_embedding_layer(key)
```

Does your output match with the original model after the weight transfer without doing this? (initially, I thought it's the same thing, but the outputs are different for Mistral when I don't do this)

Reply:
Oh, I see: https://github.com/huggingface/transformers/blob/ced9fd86f55ebb6b656c273f6e23f8ba50652f83/src/transformers/models/llama/modeling_llama.py#L113

Seems like the Huggingface implementation is equivalent to the `RotaryEmbedding` layer. I was looking at the [facebookresearch/llama](https://github.com/facebookresearch/llama/blob/main/llama/model.py) repo which has slighlt different implementation (which is the one that mistral uses). I will try to play around with weights to get the outputs to match on Mistral's side.